compiler/src/main/kotlin/io/spine/protodata/backend/AstComponents.kt

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright 2023, TeamDev. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.protodata.backend
+
+import com.google.protobuf.Descriptors
+import com.google.protobuf.Empty
+import io.spine.protodata.EnumConstant
+import io.spine.protodata.Field
+import io.spine.protodata.Rpc
+import io.spine.protodata.ServiceName
+import io.spine.protodata.TypeName
+import io.spine.protodata.cardinality
+import io.spine.protodata.constantName
+import io.spine.protodata.file
+import io.spine.protodata.name
+import io.spine.protodata.path
+
+/**
+ * Converts this field descriptor into a [Field] with options.
+ *
+ * @see buildField
+ */
+internal fun Descriptors.FieldDescriptor.buildFieldWithOptions(
+    declaringType: TypeName,
+    documentation: Documentation
+): Field {
+    val field = buildField(declaringType, documentation)
+    return field.toBuilder()
+        .addAllOption(listOptions(options))
+        .build()
+}
+
+/**
+ * Converts this field descriptor into a [Field].
+ *
+ * The resulting [Field] will not reflect the field options.
+ *
+ * @see buildFieldWithOptions
+ */
+internal fun Descriptors.FieldDescriptor.buildField(
+    declaringType: TypeName,
+    documentation: Documentation
+): Field {
+    return Field.newBuilder()
+        .setName(name())
+        .setDeclaringType(declaringType)
+        .setNumber(number)
+        .setOrderOfDeclaration(index)
+        .assignTypeAndCardinality(this)
+        .setDoc(documentation.forField(this))
+        .build()
+}
+
+/**
+ * Assigns the field type and cardinality (`map`/`list`/`oneof_name`/`single`) to the receiver
+ * builder.
+ *
+ * @return the receiver for method chaining.
```

Comment:
I think, in Kotlin we still don't put periods in `@return` statement descriptions.